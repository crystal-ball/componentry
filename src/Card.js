import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

// Create Named Components
import generateNamed from './generate-named';
const Block = generateNamed({ name: 'card-block' });
const Footer = generateNamed({ name: 'card-footer' });
const Header = generateNamed({ name: 'card-header' });

/**
 * Card component is a simple wrapper for creating markup for card elements. Includes:
 * - `Card.Block`
 * - `Card.Header`
 * - `Card.Footer`
 */
export default class Card extends Component {
  static Block = Block
  static Header = Header
  static Footer = Footer

  render() {
    const {
      As='div',
      className='',
      children=null,
      ...other
    } = this.props;
    let _className = classnames('card', className);

    return (
      <As className={_className} {...other}>
        {children}
      </As>
    );
  }
}

Card.propTypes = {
  As: PropTypes.any,
  children: PropTypes.node,
  className: PropTypes.string,
};

Card.defaultProps = {
  As: 'div',
  children: null,
  className: '',
};
