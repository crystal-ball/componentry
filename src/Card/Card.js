// @flow
import elem from '../elem-factory'
import withTheme from '../withTheme'

/**
 * Card component is a simple wrapper for creating markup for card elements
 */
const Card = withTheme('Card', props => elem({ classes: 'card', ...props }))

const Body = withTheme('CardBody', props => elem({ classes: 'card-body', ...props }))
Card.Body = Body

const Footer = withTheme('CardFooter', props =>
  elem({ classes: 'card-footer', ...props }),
)
Card.Footer = Footer

const Header = withTheme('CardHeader', props =>
  elem({ classes: 'card-header', ...props }),
)
Card.Header = Header

const Title = withTheme('CardTitle', props =>
  elem({ defaultAs: 'h4', classes: 'card-title', ...props }),
)
Card.Title = Title

export default Card
