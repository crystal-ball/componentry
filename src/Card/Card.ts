/* eslint-disable @typescript-eslint/no-empty-interface */
import React from 'react'
import { ComponentBaseProps } from '../utils/types'
import { staticComponent } from '../utils/static-component-builder'

interface CardProps extends ComponentBaseProps<'div'> {}
interface CardBodyProps extends ComponentBaseProps<'div'> {}
interface CardFooterProps extends ComponentBaseProps<'div'> {}
interface CardHeaderProps extends ComponentBaseProps<'div'> {}
interface CardTitleProps extends ComponentBaseProps<'div'> {}

export interface Card {
  (props: CardProps): React.ReactElement
  displayName: 'Card'
  /**
   * [Card body component 📝](https://componentry.design/components/card)
   */
  Body: React.FC<CardBodyProps>
  /**
   * [Card footer component 📝](https://componentry.design/components/card)
   */
  Footer: React.FC<CardFooterProps>
  /**
   * [Card header component 📝](https://componentry.design/components/card)
   */
  Header: React.FC<CardHeaderProps>
  /**
   * [Card title component 📝](https://componentry.design/components/card)
   */
  Title: React.FC<CardTitleProps>
}

/**
 * [Card component 📝](https://componentry.design/components/card)
 * @experimental
 */
export const Card = staticComponent('Card') as Card

Card.Body = staticComponent('CardBody')

Card.Footer = staticComponent('CardFooter')

Card.Header = staticComponent('CardHeader')

Card.Title = staticComponent('CardTitle', {
  as: 'h4',
})
