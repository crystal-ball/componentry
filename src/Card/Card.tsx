import React from 'react'
import { staticComponent } from '../factories/static-component'
import { BaseProps } from '../utils/base-types'

interface CardProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}
interface CardBodyProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}
interface CardFooterProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}
interface CardHeaderProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}
interface CardTitleProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface Card {
  (props: CardProps): React.ReactElement
  displayName: 'Card'
  /**
   * [Card body component 📝](https://componentry.design/components/card)
   */
  Body: React.FC<CardBodyProps>
  /**
   * [Card footer component 📝](https://componentry.design/components/card)
   */
  Footer: React.FC<CardFooterProps>
  /**
   * [Card header component 📝](https://componentry.design/components/card)
   */
  Header: React.FC<CardHeaderProps>
  /**
   * [Card title component 📝](https://componentry.design/components/card)
   */
  Title: React.FC<CardTitleProps>
}

/** [Card component 📝](https://componentry.design/components/card) */
export const Card = staticComponent('Card', {
  componentCx: 'card',
}) as Card
Card.displayName = 'Card'

Card.Body = staticComponent<CardBodyProps>('CardBody', {
  componentCx: 'card-body',
})

Card.Footer = staticComponent<CardFooterProps>('CardFooter', {
  componentCx: 'card-footer',
})

Card.Header = staticComponent<CardHeaderProps>('CardHeader', {
  componentCx: 'card-header',
})

Card.Title = staticComponent<CardTitleProps>('CardTitle', {
  as: 'h4',
  componentCx: 'card-title',
})
