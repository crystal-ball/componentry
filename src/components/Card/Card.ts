import React from 'react'
import { useTheme } from '../Theme/Theme'
import { element } from '../../utils/element-creator'
import { staticComponent } from '../../utils/static-component-builder'
import { ComponentBaseProps } from '../../utils/types'

interface CardProps extends ComponentBaseProps<'div'> {
  variant?: 'outlined'
}
interface CardBodyProps extends ComponentBaseProps<'div'> {}
interface CardFooterProps extends ComponentBaseProps<'div'> {}
interface CardHeaderProps extends ComponentBaseProps<'div'> {}
interface CardTitleProps extends ComponentBaseProps<'h4'> {}
interface CardSubtitleProps extends ComponentBaseProps<'h5'> {}

export interface Card {
  (props: CardProps): React.ReactElement
  displayName: 'Card'
  /**
   * [Card body component 📝](https://componentry.design/components/card)
   */
  Body: React.FC<CardBodyProps>
  /**
   * [Card footer component 📝](https://componentry.design/components/card)
   */
  Footer: React.FC<CardFooterProps>
  /**
   * [Card header component 📝](https://componentry.design/components/card)
   */
  Header: React.FC<CardHeaderProps>
  /**
   * [Card subtitle component 📝](https://componentry.design/components/card)
   */
  Subtitle: React.FC<CardSubtitleProps>
  /**
   * [Card title component 📝](https://componentry.design/components/card)
   */
  Title: React.FC<CardTitleProps>
}

/**
 * [Card component 📝](https://componentry.design/components/card)
 * @experimental
 */
export const Card = ((props) => {
  const { variant = 'outlined', ...rest } = {
    ...useTheme<CardProps>('Card', props.__precompile),
    ...props,
  }

  return element({
    componentCx: `🅲Card-base 🅲Card-${variant}`,
    ...rest,
  })
}) as Card
Card.displayName = 'Card'

Card.Body = staticComponent('CardBody')

Card.Footer = staticComponent('CardFooter')

Card.Header = staticComponent('CardHeader')

Card.Title = staticComponent<CardTitleProps>('CardTitle', {
  as: 'h4',
})

Card.Subtitle = staticComponent<CardSubtitleProps>('CardSubtitle', {
  as: 'h5',
})
