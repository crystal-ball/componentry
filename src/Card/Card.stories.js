/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'

import Card from './Card'
import Anchor from '../Anchor/Anchor'

storiesOf('Card', module).add('<Card />', () => (
  <div className='d-flex flex-column'>
    <Card maxWidth='50%'>
      <Card.Header>Card Header</Card.Header>
      <Card.Body>
        <Card.Title>Card title</Card.Title>
        <h5 className='card-subtitle'>Card subtitle</h5>
        <p className='card-text'>
          Card.Body example content. Note: If content with the <code>card-text</code>{' '}
          class is the last child of the body the margin will be set to 0 for whitespace
          consistency
        </p>

        <p className='card-text'>
          <Anchor className='card-link' href='#'>
            Card link
          </Anchor>
          <Anchor className='card-link' href='#'>
            Another link
          </Anchor>
        </p>
      </Card.Body>
      <Card.Footer>2 days ago</Card.Footer>
    </Card>

    <br />

    <Card maxWidth='50%'>
      <svg
        className='bd-placeholder-img card-img-top'
        width='100%'
        height='180'
        xmlns='http://www.w3.org/2000/svg'
        preserveAspectRatio='xMidYMid slice'
        focusable='false'
        role='img'
        aria-label='Placeholder: Image cap'
      >
        <title>Placeholder</title>
        <rect width='100%' height='100%' fill='#fef' />
        <text x='50%' y='50%' fill='#6c757d' dx='-38px' dy='8px'>
          Image cap
        </text>
      </svg>
      {/* <Card.Header>Card Header</Card.Header> */}
      <Card.Body>
        <Card.Title>Card title</Card.Title>
        <h5 className='card-subtitle'>Card subtitle</h5>
        <p className='card-text'>
          Card.Body example content. Note: If content with the <code>card-text</code>{' '}
          class is the last child of the body the margin will be set to 0 for whitespace
          consistency
        </p>

        <p className='card-text'>
          <Anchor className='card-link' href='#'>
            Card link
          </Anchor>
          <Anchor className='card-link' href='#'>
            Another link
          </Anchor>
        </p>
      </Card.Body>
    </Card>
  </div>
))
