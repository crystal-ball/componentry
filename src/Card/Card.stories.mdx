import { Meta } from '@storybook/addon-docs'
import { BADGE } from '@geometricpanda/storybook-addon-badges'

import { Link } from '../Link/Link'
import { List } from '../List/List'
import { Card } from './Card'

<Meta
  title='Components/Card'
  component={Card}
  parameters={{ badges: [BADGE.EXPERIMENTAL] }}
/>

# Card

<Card maxWidth='50%'>
  <Card.Header>Card Header</Card.Header>
  <Card.Body>
    <Card.Title>Card title</Card.Title>
    <h5 className='card-subtitle'>Card subtitle</h5>
    <p className='card-text'>
      Card.Body example content. Note: If content with the <code>card-text</code> class is
      the last child of the body the margin will be set to 0 for whitespace consistency
    </p>
    <p className='card-text'>
      <Link className='card-link' href='#demo'>
        Card link
      </Link>
      <Link className='card-link' href='#demo'>
        Another link
      </Link>
    </p>
  </Card.Body>
  <Card.Footer>2 days ago</Card.Footer>
</Card>

## Card elements

<Card width='50%'>
  <svg
    className='bd-placeholder-img card-img-top'
    width='100%'
    height='180'
    xmlns='http://www.w3.org/2000/svg'
    preserveAspectRatio='xMidYMid slice'
    focusable='false'
    role='img'
    aria-label='Placeholder: Image cap'
  >
    <title>Placeholder</title>
    <rect width='100%' height='100%' fill='#fef' />
    <text x='50%' y='50%' fill='#6c757d' dx='-38px' dy='8px'>
      Image cap
    </text>
  </svg>
  <Card.Body>
    <Card.Title>Card title</Card.Title>
    <h5 className='card-subtitle'>Card subtitle</h5>
    <p className='card-text'>
      Must go faster... go, go, go, go, go! I was part of something special. This thing
      comes fully loaded. AM/FM radio, reclining bucket seats, and... power windows.
      You&apos;re a very talented young man, with your own clever thoughts and ideas. Do
      you need a manager?
    </p>
  </Card.Body>
  <List flush>
    <List.Item>God creates dinosaurs.</List.Item>
    <List.Item>God destroys dinosaurs.</List.Item>
    <List.Item>God creates Man.</List.Item>
    <List.Item>Man destroys God.</List.Item>
    <List.Item>Man creates Dinosaurs.</List.Item>
  </List>
  <Card.Body>
    <a href='#demo' className='card-link'>
      Card link
    </a>
    <a href='#demo' className='card-link'>
      Another link
    </a>
  </Card.Body>
  <Card.Footer>God help us, we&apos;re in the hands of engineers.</Card.Footer>
</Card>
