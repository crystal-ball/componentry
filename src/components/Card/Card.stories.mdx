import { Meta, Story } from '@storybook/addon-docs'

import { Flex } from '../Flex/Flex'
import { Link } from '../Link/Link'
import { Text } from '../Text/Text'
import { Card } from './Card'

<Meta title='Components/Card' component={Card} />

# Card

<Story name='Card'>
  <Card maxWidth='50%'>
    <Card.Header>
      <Text>Card Header</Text>
    </Card.Header>
    <Card.Body>
      <Card.Title>Card title</Card.Title>
      <Card.Subtitle>Card subtitle</Card.Subtitle>
      <Text mt={2}>
        God help us, we're in the hands of engineers. So you two dig up, dig up dinosaurs?
        Checkmate... Do you have any idea how long it takes those cups to decompose. God
        help us, we're in the hands of engineers. You really think you can fly that thing?
        They're using our own satellites against us. And the clock is ticking.
      </Text>
      <Flex gap={2} mt={2}>
        <Link href='#demo'>Card link</Link>
        <Link href='#demo'>Another link</Link>
      </Flex>
    </Card.Body>
    <Card.Footer>
      <Text>2 days ago</Text>
    </Card.Footer>
  </Card>
</Story>

## Card elements

<Card width='50%'>
  <svg
    className='rounded-t'
    width='100%'
    height='180'
    xmlns='http://www.w3.org/2000/svg'
    preserveAspectRatio='xMidYMid slice'
    focusable='false'
    role='img'
    aria-label='Placeholder: Image cap'
  >
    <title>Placeholder</title>
    <rect width='100%' height='100%' fill='#fef' />
    <text x='50%' y='50%' fill='#6c757d' dx='-38px' dy='8px'>
      Image cap
    </text>
  </svg>
  <Card.Body>
    <Card.Title>Card title</Card.Title>
    <Card.Subtitle>Card subtitle</Card.Subtitle>
    <Text mt={2}>
      God help us, we're in the hands of engineers. So you two dig up, dig up dinosaurs?
      Checkmate... Do you have any idea how long it takes those cups to decompose. God
      help us, we're in the hands of engineers. You really think you can fly that thing?
      They're using our own satellites against us. And the clock is ticking.
    </Text>
  </Card.Body>
  <Card.Body display='flex' gap={2}>
    <Link href='#demo'>Card link</Link>
    <Link href='#demo'>Another link</Link>
  </Card.Body>
  <Card.Footer>God help us, we&apos;re in the hands of engineers.</Card.Footer>
</Card>
