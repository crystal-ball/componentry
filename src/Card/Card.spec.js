import React from 'react'
import { render } from '@testing-library/react'

import elementTests from '../../test/element-tests'
import Card from './Card'

describe('<Card />', () => {
  // Basic library element test suite
  elementTests(Card)
  elementTests(Card.Header)
  elementTests(Card.Body)
  elementTests(Card.Title)
  elementTests(Card.Footer)

  test('should render a container div with class card by default', () => {
    const { container } = render(<Card />)
    expect(container.firstChild).toHaveClass('card')
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Card /> snapshots', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Card>
        <Card.Header>Header</Card.Header>
        <Card.Body>
          <Card.Title>Title</Card.Title>
          Body
        </Card.Body>
        <Card.Footer>Footer</Card.Footer>
      </Card>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
