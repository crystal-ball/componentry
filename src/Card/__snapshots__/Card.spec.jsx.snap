// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> snapshots renders correctly 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    Header
  </div>
  <div
    class="card-body"
  >
    <h4
      class="card-title"
    >
      Title
    </h4>
    Body
  </div>
  <div
    class="card-footer"
  >
    Footer
  </div>
</div>
`;
